Add real-time Ubuntu Core docs (#45)
canonical/real-time-ubuntu-docs: docs/how-to/uc-boot-parameters.rst

Reviewer: Hi @farshidtz , a general comment, I would suggest adding some headings to signpost the doc. I left some suggestions through out this doc.

Thanks.

How to configure boot parameters on Ubuntu Core 
================================================

Different from classic versions of Ubuntu (Desktop and Server), `Ubuntu Core`_ is a minimal, secure and strictly confined operating system designed to run on embedded systems, such as IoT devices.
Ubuntu Core is a snap-based operating system, meaning it uses snap packages to manage applications and services.
Its read-only nature makes it a very secure and reliable operating system but also makes it a bit more difficult to configure certain aspects of the system, such as boot parameters.

> This how-to guide assumes the use of a `pre-built Ubuntu Core image`_ on a physical device rather than a virtual machine.

Author: This isn't correct. Pre-built images don't have the realtime-kernel.

The second one is used to append any other kernel boot parameters, considered as not analyzed by the gadget snap.

When using the reference gadget snaps (`pc-gadget`_ and `pi-gadget`_), you need to use the `system.kernel.dangerous-cmdline-append` option, as the allow list is not present in them.

When creating a custom `gadget snap`_ for a real-time Ubuntu Core system, it is advisable to add an allow list with the relevant parameters.

The kernel parameters in the snap configurations exactly match the `kernel parameters`_ that would be passed to the kernel on the grub command line, the :code:`GRUB_CMDLINE_LINUX_DEFAULT` on :code:`/etc/default/grub` file. 
Parameters are passed as a single string in the :code:`key=value` format, with each parameter separated by a space.

In the case of :doc:`real-time kernel parameters <../reference/kernel-boot-parameters>`, the most interesting parameters are `nohz`_, `nohz_full`_ and `irqaffinity`_.
The :code:`nohz` is used to enable/disable dynamic ticks (possible values are `on` or `off`). 
The :code:`nohz_full` receives a `cpu list`_ specifying which CPUs will have dynamic ticks disabled. 

> The `irqaffinity` is used to set the affinity of the IRQs to the CPUs, in other words: which CPUs will handle the IRQs.

Reviewer 2: I don't know if this is possible but it would be great if by hovering an accronym I would get it's definition. I had to search 'IRQs' on search bar to find it.

Author: It might be possible, but a tooltip will be hard to use on touch screen devices. I've expanded the acronym.

Reviewer: This is a little late but there is a Sphinx directive (`:abbr:`) for this exact use.
https://github.com/canonical/sphinx-docs-starter-pack/blob/main/doc-cheat-sheet.rst?plain=1#L243

But unfortunately as Farshid mentioned, this doesn't work for mobile devices / touch screens.

For Kernel docs we use a Glossary to help address this so we can direct users to a complete list of terms (including acronyms + their definitions) so we don't have to worry about expanding the acronym on each occurrence.

How to configure boot parameters on Ubuntu Core 
================================================

.. note::

    This guide assumes access to features available exclusively to `dedicated Snap Store`_ users.

Ubuntu Core doesn't have the :code:`/etc/default/grub` file mapped for the kernel parameters, neither the `update-grub`_ utility.
On Ubuntu Core, the kernel parameters can be modified statically or dynamically.
Static configurations should be added during the image build; refer to :doc:`../how-to/uc-image-creation`.
Dynamic configurations can be set `using snap options`_ on a running system.
This guide focuses on the dynamic configuration method, suitable for iteratively tuning the operating system.

.. Refer to :doc:`../how-to/uc-image-creation` if you want to create a real-time Ubuntu Core image.

> Open a terminal and access your real-time Ubuntu Core instance.

Author: I think we also need documentation on creating a real-time Ubuntu core image without static configurations, to allow iterative tuning. There are currently two how-tos:
- This one (`docs/how-to/uc-boot-parameters.rst`); assuming a real-time Ubuntu Core instance is available.
- The image creation guide (`docs/how-to/uc-image-creation.rst`), explaining how to create the image, but only after customizing a gadget to set static config.

Author: PR: https://github.com/canonical/real-time-ubuntu-docs/pull/56

How to configure boot parameters on Ubuntu Core 
================================================

.. note::

    This guide assumes access to features available exclusively to `dedicated Snap Store`_ users.

Ubuntu Core doesn't have the :code:`/etc/default/grub` file mapped for the kernel parameters, nor the `update-grub`_ utility.
On Ubuntu Core, the kernel parameters can be modified statically or dynamically.
Static configurations are added during the image build.

> On the other hand, dynamic configurations are set `using snap options`_ on a running system.

Reviewer: ```suggestion
On the other hand, dynamic configurations are set using `snap options`_ on a running system.
```

Nitpick: the link text could be more specific.

> This guide focuses on the dynamic configuration method, suitable for iteratively tuning the operating system.

Reviewer: ```suggestion

This guide focuses on the dynamic method for configuring kernel boot parameters, suitable for iteratively tuning the operating system.
```

Suggesting a line break and complete context here for readability. You can easily skim the objective of this how-to guide from that one line.

Reviewer: ![image](https://github.com/user-attachments/assets/aea7cedf-e1c1-4d7b-8233-3c71e3bcdc14)

So it looks like this.

.. hint::
    
    Refer to :doc:`../how-to/uc-image-creation` if you want to create a real-time Ubuntu Core image.

Open a terminal and access your real-time Ubuntu Core instance.
The instance should be on a physical device rather than a virtual machine.

> While a virtual machine can be used to follow the steps in this guide, it will not provide the benefits of a real-time kernel.

Reviewer: Hi @farshidtz , not sure if I'm misunderstanding the intent of this line because the VM won't be 'providing' the benefits but won't be able to 'inherit' the benefits.

```suggestion
While a virtual machine can be used to follow the steps in this guide, it will not benefit from the features of a real-time kernel.
```

Author: I think the existing explanation is ok. Running real-time workload in a VM involves the host OS, hypervisor, and the VM environment. I'll leave this for @locnnil to decide, since this text was written by him in #39.

Reviewer 3: I like @AnneCYH suggestion! That's what I meant to say, it could be used  "characteristics" as well. Applied on https://github.com/canonical/real-time-ubuntu-docs/pull/45/commits/1d8d02797da0424c35b335a1fd758a6a8f5d4a20. Thanks!

> 

Reviewer: ```suggestion
## Dynamic kernel boot parameters
```

There are two snap options for setting the kernel parameters:

- `system.kernel.cmdline-append`_
- `system.kernel.dangerous-cmdline-append`_

> The first one is used to append permitted kernel boot parameters that are verified against an `allow list`_ in the `gadget snap`_. 

Reviewer: Maybe move this as a 'description' directly under the bullet point for each snap option?
E.g.

- `system.kernel.cmdline-append` - Used to append permitted kernel boot parameters that are verified...
- `system.kernel.dangerous-cmdline-append` - Used to append any other...

So the context is present instead of being referred to as "first" or "second one"?

Thanks.

The second one is used to append any other kernel boot parameters, considered as not analyzed by the gadget snap.

When using the reference gadget snaps (`pc-gadget`_ and `pi-gadget`_), you need to use the `system.kernel.dangerous-cmdline-append` option, as the allow list is not present in them.

When creating a custom `gadget snap`_ for a real-time Ubuntu Core system, it is advisable to add an allow list with the relevant parameters.

> The kernel parameters in the snap configurations exactly match the `kernel parameters`_ that would be passed to the kernel on the grub command line, the :code:`GRUB_CMDLINE_LINUX_DEFAULT` on :code:`/etc/default/grub` file. 

Reviewer: ```suggestion
The kernel parameters in the snap configurations exactly match the `kernel parameters`_ that would be passed to the kernel on the grub command line, :code:`GRUB_CMDLINE_LINUX_DEFAULT` in the :file:`/etc/default/grub` file. 
```

Parameters are passed as a single string in the :code:`key=value` format, with each parameter separated by a space.

In the case of :doc:`real-time kernel parameters <../reference/kernel-boot-parameters>`, the most interesting parameters are `nohz`_, `nohz_full`_ and `irqaffinity`_.

> The :code:`nohz` is used to enable/disable dynamic ticks (possible values are `on` or `off`). 

Reviewer: This is rather a nitpick also.
But I would suggest some edits for readability similar to this:
https://documentation.ubuntu.com/real-time/en/latest/how-to/tune-irq-affinity/#how-to-tune-irq-affinity

```
...the most interesting parameters are `nohz`_, `nohz_full`_ and `irqaffinity`_.
- :code:`nohz` - Used to enable...
- :code:`nohz_full` - Receives...
```

The :code:`nohz_full` receives a `cpu list`_ specifying which CPUs will have dynamic ticks disabled. 

> The `irqaffinity` is used to set the affinity of the interrupt requests (IRQs) to the CPUs, in other words: which CPUs will handle the IRQs.

Reviewer: ```suggestion
The :code:`irqaffinity` is used to set the affinity of the interrupt requests (IRQs) to the CPUs, in other words: which CPUs will handle the IRQs.
```

> For example:

Reviewer: I know this is not a tutorial but it might be useful to help the user confirm the command is doing what they think it's doing with a brief summary.

```
For example, the snap command below enables dynamic ticks for at least 2 CPUs...
```

Author: @locnnil do you agree with https://github.com/canonical/real-time-ubuntu-docs/pull/45/commits/6db6a330989c8a7b9bd775476e79c26ea3876519 ?

Reviewer 3: There are some problems with the statement on https://github.com/canonical/real-time-ubuntu-docs/commit/6db6a330989c8a7b9bd775476e79c26ea3876519:

- "enables dynamic ticks" : That's true, but you may want to add a URL on that, or just add on parenthesis: 
"(avoid sending scheduling-clock interrupts to CPUs with a single runnable task)" This is also referred to as `adaptive-ticks` in other places on the kernel, both have the same meaning, I would prefer to use `adaptive-ticks`, but I leave it up to you. You can check [this kernel doc snippet](https://github.com/torvalds/linux/blob/3ec3f5fc4a91e389ea56b111a73d97ffc94f19c6/Documentation/timers/no_hz.rst?plain=1#L112C1-L117C43) with a precise explanation on what NO_HZ is about.

- "disabling them for all CPUs 2-N" : This is wrong, it's not disabling `dynamic ticks`/`adaptive-ticks`, it is enabling it for the selected CPUs, [as you can check here](https://github.com/torvalds/linux/blob/3ec3f5fc4a91e389ea56b111a73d97ffc94f19c6/Documentation/timers/no_hz.rst?plain=1#L124-L125).

The final part of the sentence is corret.

Author: @locnnil 
> "disabling them for all CPUs 2-N" : This is wrong, it's not disabling dynamic ticks/adaptive-ticks, it is enabling it for the selected CPUs, [as you can check here](https://github.com/torvalds/linux/blob/3ec3f5fc4a91e389ea56b111a73d97ffc94f19c6/Documentation/timers/no_hz.rst?plain=1#L124-L125).

So the explanation above it is wrong, coming from #39.

> For example:

Reviewer: ```suggestion
## Modify kernel boot parameters

Use the :code:`snap set` command to dynamically modify the kernel parameters.
For example...
```

Author: Is this still useful after the refactoring in https://github.com/canonical/real-time-ubuntu-docs/pull/45/commits/e4f0a1de448f4c372b7a500ca70d38bada4b41f4?

The sections are rather short now.

.. code-block:: bash
    
    snap set system system.kernel.dangerous-cmdline-append="nohz=on nohz_full=2-N irqaffinity=0-1"

After setting the kernel parameters, it's necessary to reboot the system to apply the changes.
Then you can check if the parameters were applied:

> .. code-block:: console

Reviewer: We actually have a terminal view directive which I personally like.
![image](https://github.com/user-attachments/assets/f2e852f3-9988-4d07-90dd-42008ac014c0)

It just helps distinguish sections where we want to direct user to the expected output.

```
.. terminal::
   :input: cat /proc/cmdline
   :user: user
   :host: host
   :dir: ~

   snapd_recovery_mode=run console=ttyS0,115200n8 console=tty1 panic=-1 nohz=on nohz_full=2-N irqaffinity=0-1
```

Author: I'm aware of this but it has a completely different style than the normal code blocks. It doesn't look good when used together with code blocks on a single document.

    $ cat /proc/cmdline
    snapd_recovery_mode=run console=ttyS0,115200n8 console=tty1 panic=-1 nohz=on nohz_full=2-N irqaffinity=0-1

To undo the modifications made to the kernel parameters, use the `snap unset`_ command:

.. code-block:: bash

    snap unset system system.kernel.dangerous-cmdline-append

> ---

Reviewer: Is the line here intentional?
Thanks.

Author: The intention was to conclude, separate from that section of the document.

.. _update-grub: https://manpages.ubuntu.com/manpages/xenial/man8/update-grub.8.html
.. _snap set: https://ubuntu.com/core/docs/modify-kernel-options
.. _system.kernel.cmdline-append: https://snapcraft.io/docs/system-options#heading--kernel-cmdline-append
.. _system.kernel.dangerous-cmdline-append: https://snapcraft.io/docs/system-options#heading--kernel-dangerous-cmdline-append
.. _gadget snap: https://snapcraft.io/docs/the-gadget-snap
.. _allow list: https://snapcraft.io/docs/the-gadget-snap#heading--dynamic
.. _nohz: https://docs.kernel.org/timers/no_hz.html
.. _nohz_full: https://docs.kernel.org/timers/no_hz.html#omit-scheduling-clock-ticks-for-cpus-with-only-one-runnable-task
.. _kernel parameters: https://docs.kernel.org/admin-guide/kernel-parameters.html
.. _cpu list: https://docs.kernel.org/admin-guide/kernel-parameters.html#cpu-lists
.. _irqaffinity: https://docs.kernel.org/core-api/irq/irq-affinity.html
.. _snap unset: https://snapcraft.io/docs/set-system-options
.. _building a gadget snap: https://ubuntu.com/core/docs/gadget-building
.. _pc-gadget: https://github.com/snapcore/pc-gadget
.. _pi-gadget: https://github.com/snapcore/pi-gadget

> .. _using snap options: https://snapcraft.io/docs/system-options

Reviewer: ```suggestion
.. _ snap options: https://snapcraft.io/docs/system-options
```

Based on earlier comment on link text since this goes to the list of snap options.
